### PR TITLE
Add excludedIds prop to DatagridOverlay containers

### PR DESCRIPTION
**Pull request by nnatter**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| License | MIT

#### What's in this PR?

This PR add a `excludedIds` prop to the `SingleDatagridOverlay` and the `MultiDatagridOverlay`that allow to set an array of ids that are sent to the server as `excluded` query-parameter.

#### Why?

Because excluding specific ids from a `DatagridOverlay` might be handy in various situations. A simple example would be a `DatagridOverlay` that is used to add items to some list and should exclude items which are already added to that list.

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/MultiDatagridOverlay/MultiDatagridOverlay.js [hunk 1]
```diff
@@ -1,7 +1,9 @@
 // @flow
 import React from 'react';
-import {observable} from 'mobx';
+import {observable, autorun, action, toJS} from 'mobx';
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/MultiDatagridOverlay/MultiDatagridOverlay.js]: alphabet :slightly_smiling_face:

**Comment by nnatter** [comment]
@danrot the PR now adjusts our `MultiSelection` container to use the `excludedIds` query-parameter to exclude items which are already selected from the `DatagridOverlay`.

File: src/Sulu/Component/Rest/ListBuilder/AbstractListBuilder.php [hunk 2]
```diff
@@ -59,6 +59,21 @@ abstract class AbstractListBuilder implements ListBuilderInterface
...
+     * If null, the result is not restricted to specific ids.
+     *
+     * @var ?array
+     */
+    protected $ids = null;
+
+    /**
+     * Array of ids which are excluded from the result of the query.
+     *
+     * @var array
```
[the remark this answers is not in the document]
Comment (danrot) [reply src/Sulu/Component/Rest/ListBuilder/AbstractListBuilder.php]: I'd say so. Was quite confusing for me :see_no_evil:

[the remark this answers is not in the document]
Comment (nnatter) [reply src/Sulu/Component/Rest/ListBuilder/AbstractListBuilder.php]: done 🙂

File: src/Sulu/Component/Rest/ListBuilder/Doctrine/DoctrineListBuilder.php [hunk 3]
```diff
@@ -526,6 +526,16 @@ protected function createQueryBuilder($joins = null)
 
         $this->assignJoins($this->queryBuilder, $joins);
 
+        if (null !== $this->ids) {
+            $this->queryBuilder->andWhere($this->idField->getSelect() . ' in (:idsFilter)');
```
[the remark this answers is not in the document]
Comment (nnatter) [reply src/Sulu/Component/Rest/ListBuilder/Doctrine/DoctrineListBuilder.php]: changed these calls to use the `in` and `notIn` method of the list-builder.

**Comment by danrot** [comment]
This does not work correctly for Snippets anymore. The item is now neither excluded nor selected when opening the overlay.

**Comment by nnatter** [comment]
This PR currently lacks the handling of the `ids` and `exludedIds` parameter
 in the `SnippetController` and the `PageController`.

**Comment by nnatter** [comment]
@danrot after talking to @chirimoya we decided to revert the changes to our `MultiSelection` container/`Selection` field type. 

We cannot exclude already selected items from the `ListOverlay` of the `MultiSelection` because the items might be presented in a tree in the `ListOverlay` (eg. pages or categories). This would lead to undesired/strange behaviour when excluding an item that has child-items.

**Comment by danrot** [comment]
@nnatter That reasoning sounds right. However, what is this PR really doing then? Does that mean that in none of the selection field types the selected items should be excluded? The changes look to me this way, and my question is if the media selection shouldn't behave the same way then. Because currently this field type is the only one left excuding the selected items.

**Comment by nnatter** [comment]
@danrot The PR implements a `excludedIds` prop on the `ListOverlay` containers that allows to exclude items from the overlay. For example, this is useful when implementing a custom component that uses our `ListOverlay` to add some items (that are not removable afterwards) and you want to exclude already added items from the overlay.
Furthermore, the PR implements the handling of the `ids` and `excludedIds` query parameter in our `ListRestHelper` and `ListBuilder`. This allows to simplify some controllers that implemented this logic on their own before.

As the `excludedIds` prop of the `ListOverlay` containers is not used anywhere in the core right now, this PR should not change any behaviour in its current state.

**Comment by nnatter** [comment]
Regarding to our `MediaSelection` containers: Yes their behaviour is not consistent to the other `Selection` containers right now (as stated above, this PR actually does not change any behaviour). 

I think it would be better if the behaviour is consistent here too. But i do not know if there is a reason for this inconsistency - do you? @danrot @chirimoya 

Anyhow, i would prefer to make these things consistent in a separate PR as that matter is not directly associated to this PR in my opinion. Also, we could remove on fork from our project 😉 Is that okay for you?

File: UPGRADE.md [hunk 4]
```diff
@@ -2,6 +2,10 @@
 
 ## dev-develop
 
+### excluded query parameter of Media API
+The `excluded` query parameter which can be used to exclude specific ids from the media list returned by the Media API
```
Comment (danrot) [inline UPGRADE.md]: We usually have a new line after the header first :slightly_smiling_face:

**Comment by danrot** [comment]
Ok, if the behavior is really exactly the same as before it is fine for me. Can you still create an issue regarding the media selection and add it to the Release 2.0 milestone?

**Pull request merged**